volcli/volcli.go

```diff
@@ -286,6 +289,123 @@ func policyList(ctx *cli.Context) (bool, error) {
 	return false, nil
 }
 
+// PolicyGetRevision retrieves a single revision from a policy's history.
+func PolicyGetRevision(ctx *cli.Context) {
+	execCliAndExit(ctx, policyGetRevision)
+}
+
+func policyGetRevision(ctx *cli.Context) (bool, error) {
+	if len(ctx.Args()) != 2 {
+		return true, errorInvalidArgCount(len(ctx.Args()), 2, ctx.Args())
+	}
+
+	name := ctx.Args()[0]
+	revision := ctx.Args()[1]
+
+	resp, err := http.Get(fmt.Sprintf("http://%s/policy-archives/%s/%s",
+		ctx.GlobalString("apiserver"),
+		name,
+		revision,
+	))
+	if err != nil {
+		return false, err
+	}
+
+	if resp.StatusCode != 200 {
+		if _, err := io.Copy(os.Stderr, resp.Body); err != nil {
+			return false, errored.Errorf("Error copying body: %v\nResponse Status Code was %d, not 200", err, resp.StatusCode)
+		}
+		return false, errored.Errorf("Response Status Code was %d, not 200", resp.StatusCode)
+	}
+
+	content, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, err
+	}
+
+	fmt.Println(string(content))
+
+	return false, nil
+}
+
+// PolicyListRevisions retrieves all the revisions for a given policy.
+func PolicyListRevisions(ctx *cli.Context) {
+	execCliAndExit(ctx, policyListRevisions)
+}
+
+func policyListRevisions(ctx *cli.Context) (bool, error) {
+	var revisions []string
+
+	if len(ctx.Args()) != 1 {
+		return true, errorInvalidArgCount(len(ctx.Args()), 1, ctx.Args())
+	}
+
+	name := ctx.Args()[0]
+
+	resp, err := http.Get(fmt.Sprintf("http://%s/policy-archives/%s",
+		ctx.GlobalString("apiserver"),
+		name,
+	))
+	if err != nil {
+		return false, err
+	}
+
+	if resp.StatusCode != 200 {
+		if _, err := io.Copy(os.Stderr, resp.Body); err != nil {
+			return false, errored.Errorf("Error copying body: %v\nResponse Status Code was %d, not 200", err, resp.StatusCode)
+		}
+		return false, errored.Errorf("Response Status Code was %d, not 200", resp.StatusCode)
+	}
+
+	content, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, err
+	}
+
+	if err := json.Unmarshal(content, &revisions); err != nil {
+		return false, err
+	}
+
+	for _, revision := range revisions {
+		fmt.Println(revision)
+	}
+
+	return false, nil
+}
+
+// PolicyWatch watches etcd for policy changes and prints them.
+func PolicyWatch(ctx *cli.Context) {
+	execCliAndExit(ctx, policyWatch)
+}
+
+func policyWatch(ctx *cli.Context) (bool, error) {
+	if len(ctx.Args()) != 0 {
+		return true, errorInvalidArgCount(len(ctx.Args()), 0, ctx.Args())
+	}
+
+	cfg, err := config.NewClient(ctx.GlobalString("prefix"), ctx.GlobalStringSlice("etcd"))
+	if err != nil {
+		return false, err
+	}
+
+	wc := make(chan *watch.Watch)
+
+	cfg.WatchForPolicyChanges(wc, func(resp *client.Response, w *watch.Watcher) {
+		w.Channel <- &watch.Watch{Key: resp.Node.Key, Config: true}
+	})
+
+	for w := range wc {
+		keyspace, revision := path.Split(w.Key)
+		keyspace = keyspace[:len(keyspace)-1] // remove trailing slash
+
+		_, name := path.Split(keyspace)
```

Comment:
Looks like we don't actually support uploading policies with a slash in them (volcli gets a 404 back), but I flipped the logic here to just chop off the prefix and use `path.Split()`:  https://github.com/contiv/volplugin/pull/367/commits/0a01001e2fd093193632fd414bd07d792c0fa4c0#diff-fa43fdf657b1ed000108d8be7501ee81R70